File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        size = 0
        dummy = ListNode(0)
        dummy.next = head
        cur = head
        while cur:
            size += 1
            cur = cur.next
        
        nth_node = size - n + 1
        cycle = 0
        cur = head
        prev = dummy
        while cur:
            cycle += 1
            if cycle == nth_node:
                prev.next = cur.next
                break
            prev, cur = cur, cur.next
        return dummy.next


        

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int N = 0;
        ListNode* curr = head;
        while (curr) {
            N++;
            curr = curr->next;
        }

        int idx = N - n;
        if (idx == 0) {
            return head->next;
        }

        curr = head;
        for (int i = 0; i < N; ++i) {
            if (i + 1 == idx) {
                curr->next = curr->next->next;
                break;
            }
            curr = curr->next;
        }

        return head;
    }
};